packages/cli/src/rollup-plugins/server-components.ts

```diff
@@ -0,0 +1,75 @@
+import { AcornNode, Plugin, SourceDescription } from "rollup";
+import { Program } from "estree";
+import MagicString, { SourceMap } from "magic-string";
+
+export function serverComponentsPlugin({
+  sourceMap,
+}: {
+  sourceMap: boolean;
+}): Plugin {
+  return {
+    name: "server-components",
+
+    transform(code, id) {
+      if (code.includes("'use client'") || code.includes('"use client"')) {
+        const ast: Program = (() => {
+          const babelMeta = this.getModuleInfo(id)!.meta.babel;
+          if (babelMeta?.codeAtBabelTime === code) {
+            return this.getModuleInfo(id)!.meta.babel.ast;
+          }
+          return this.parse(code);
+        })();
+
+        for (const [idx, node] of ast.body.entries()) {
+          if (
+            node.type !== "ExpressionStatement" ||
+            node.expression.type !== "Literal" ||
+            typeof node.expression.value !== "string"
+          ) {
+            return;
+          }
+          if (node.expression.value === "use client") {
+            let magicString = new MagicString(code);
+            this.emitFile({ type: "chunk", id });
+            ast.body.splice(idx, 1);
+            const start: number = (node as any).start;
+            const end: number = (node as any).end;
+            const len = end - start;
+            magicString.overwrite(start, end, " ".repeat(len));
+            let output: SourceDescription = {
+              code: magicString.toString(),
+              ast: (ast as unknown) as AcornNode,
+            };
+
+            if (sourceMap) {
+              output.map = magicString.generateMap({ hires: true });
+            }
+            if (!output.meta) {
+              output.meta = {};
+            }
+            output.meta.isUseClientEntry = true;
+            return output;
+          }
+        }
+      }
+      return null;
+    },
+    renderChunk(code, chunk) {
+      if (chunk.facadeModuleId !== null) {
+        const moduleInfo = this.getModuleInfo(chunk.facadeModuleId);
+        if (moduleInfo?.meta.isUseClientEntry) {
+          const magicString = new MagicString(code);
+          magicString.prepend('"use client";\n');
```

Comment:
i guess this is because Rollup might drop the directive, right?

Reply:
Rollup actually throws a warning if you have directives so this removes the directive and then adds it back.

Reply:
Do u know if that's also the case in Rollup 3 (IIRC we are still on Rollup 2 here)?

Reply:
It is - https://github.com/rollup/rollup/blob/e0d9681de58cee80f162d02c65de293954c4a13f/src/ast/nodes/ExpressionStatement.ts#L20

Reply:
Interesting, it might be worth creating a Rollup issue about this. I suspect that more people might need to handle those directives in the near-ish future.

Reply:
tbh, imo the current behaviour is the most sensible option. There's no correct answer to bundling directives so erroring on it seems like the ideal thing.